README.md
Readme for package maic
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

# maic

<!-- badges: start -->

<!-- badges: end -->

maic is repeatable workflow for calculating subject/observation weights
for Matching Adjusted Indirect Comparison (MAIC).

MAIC is a form of population adjustment for indirect comparison that can
be undertaken when only aggregate data is available for the comparator
intervention, but full subject-level data is available for the
investigative intervention. By the method of moments, subject-level
weights are generated such that the weighted moments of the baseline
variables of the “index” intervention study match the reported moments
of the “target” study. Via the first two moments (mean and variance)
matching can be achieved on:

  - Mean of variable
  - Proportion of population with attribute
  - Proportion of population with variable below arbitrary value (by
    encoding as attribute)
  - Median or arbitrary quantile (by matching proportion of population
    below quantile value)
  - Variance of variable about matched mean
  - Standard deviation of variable about matched mean

In addition, subjects can be assigned zero weight where there is no
overlap with the target study, i.e.:

  - Subjects with attribute absent in target study (target proportion =
    0)
  - Subjects without attribute universal in target study (target
    proportion = 1)
  - Subjects with variable below minimum observed in target study
  - Subjects with variable above maximum observed in target study

The generated weights may then be used to form weighted comparisons
between intervention outcomes having removed bias for prognostic or
effect modifying covariables.

## Installation

You can install the released version of maic from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("maic")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("heorltd/maic")
```

## Example

This example demonstrates how to create the MAIC weights by creation of
the configuration objects and running the optimisation procedure:

``` r
library(maic)
target <- c("Air.Flow" = 60,
           "Water.Temp" = 21,
           "Prop.Acid.Conc.LT.90" = 0.7,
           "min.air.flow" = 55)

stackloss$match.conc.lt.90 <- 
  ifelse(stackloss$Acid.Conc. < 90, 1, 0)

dict <- data.frame(
  "match.id" = 
    c("airflow", "watertemp", 
      "acidconc", "min.airflow"),
  "target.variable" = 
    c("Air.Flow", "Water.Temp",
      "Prop.Acid.Conc.LT.90", "min.air.flow"),
  "index.variable" = 
    c("Air.Flow", "Water.Temp",
      "match.conc.lt.90", "Air.Flow"),
  "match.type" = 
    c("mean", "mean", "proportion", "min"),
  stringsAsFactors = FALSE)

ipmat <- createMAICInput(
  index = stackloss,
  target = target,
  dictionary = dict,
  matching.variables = 
    c("airflow", "watertemp", 
      "acidconc", "min.airflow"))

wts <- maicWeight(ipmat)
```

``` r
plot(wts)
```

<img src="man/figures/README-plotWeights-1.png" width="100%" />


NEWS.md
News for package maic
================
Rob Young (<rob.young@heor.co.uk>)
26/04/2022

# 26 Apr 2022

The following fixes have been made:

-   Leading zeros are handled correctly on the quantile specification,
    allowing for quantiles less than 0.1
-   Dummy variables at the quantile level are set to 0.5

# 11 May 2020

The following fixes have been made:

-   Reporting when probabilities are equal to 1 or 0 no longer fails
    (#1)
-    is now used to summarise
-   Index and dictionary inputs are forced to data.frame (#2)
-   Clarification that the p-values for difference in mean are
    stochastically derived (#3)

# 17 April 2020

This is the first CRAN upload of . Read the documentation, check the
function signatures, have a look at the attributes of what gets
returned. There are many little features that have been added to improve
your workflow, but there are major changes from the development
versions, especially:

-   Weights will now always be same same length as the number of rows in
    your data. You will no longer need to manually put in zeros for
    those subjects excluded due to lack of overlap.


tests/testthat.R
library(testthat)
library(maic)

test_check("maic")


tests/testthat/test_maicInput.R
context("MAIC input")
library(maic)

test_that("A valid input matrix can be made for individual comparisons", {
  
  # Use the datasets::stackloss dataset for these tests
  data("stackloss")
  stackloss$match.conc.lt.90 <- ifelse(stackloss$Acid.Conc. < 90, 1, 0)
  
  mtch.targets <- list(
    "Air.Flow.mean" = 60,
    "Air.Flow.min" = 55,
    "Air.Flow.max" = 90,
    "Air.Flow.sd" = 9,
    "Air.Flow.var" = 81,
    "Proportion.Acid.Conc.lt.90" = 0.7,
    "Water.Temp.decile.1" = 18,
    "Water.Temp.decile.9" = 25,
    "Water.Temp.median" = 22
  )
  
  mtch.dict <- data.frame(match.id = c(
    "airflow.mean", "airflow.min", "airflow.max", "airflow.sd", 
    "airflow.var", "acidconc.prop",
    "watertemp.10tile", "watertemp.90tile", "watertemp.median"
  ),
  match.type = c(
    "mean", "min", "max", "sd",
    "var", "proportion",
    "quantile.1", "quantile.90", "median"
  ),
  target.variable = c(
    "Air.Flow.mean", "Air.Flow.min", "Air.Flow.max", "Air.Flow.sd",
    "Air.Flow.var", "Proportion.Acid.Conc.lt.90",
    "Water.Temp.decile.1", "Water.Temp.decile.9", "Water.Temp.median"
  ),
  index.variable = c(
    "Air.Flow", "Air.Flow", "Air.Flow", "Air.Flow", 
    "Air.Flow", "match.conc.lt.90",
    "Water.Temp", "Water.Temp", "Water.Temp"
  ),
  supplementary.target.variable = c(
    "", "", "", "Air.Flow.mean",
    "Air.Flow.mean", "",
    "", "", ""
  ),
  stringsAsFactors = FALSE
  )
  
  # Mean
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean")
                           )
  expect_is(ipmat, "MaicInput")
  expect_equal(dim(ipmat$input.matrix), c(21, 1))
  expect_match(colnames(ipmat$input.matrix), "airflow.mean")
  expect_equal(min(ipmat$input.matrix[, 1]), min(stackloss$Air.Flow) - 
                 mtch.targets[["Air.Flow.mean"]])
  expect_equal(max(ipmat$input.matrix[, 1]), max(stackloss$Air.Flow) - 
                 mtch.targets[["Air.Flow.mean"]])
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 1)
  
  # Proportion
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("acidconc.prop"))
  expect_is(ipmat, "MaicInput")
  expect_equal(dim(ipmat$input.matrix), c(21, 1))
  expect_match(colnames(ipmat$input.matrix), "acidconc.prop")
  expect_equal(min(ipmat$input.matrix[, 1]), 
               min(stackloss$match.conc.lt.90) - 
                 mtch.targets[["Proportion.Acid.Conc.lt.90"]])
  expect_equal(max(ipmat$input.matrix[, 1]), 
               max(stackloss$match.conc.lt.90) - 
                 mtch.targets[["Proportion.Acid.Conc.lt.90"]])
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 1)
  
  # Minimum
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.min")
  )
  expect_is(ipmat, "MaicInput")
  expect_equal(sum(ipmat$excluded),
               sum(stackloss$Air.Flow < mtch.targets[["Air.Flow.min"]]))
  expect_equal(dim(ipmat$input.matrix), c(sum(!ipmat$excluded), 0))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 0)
  
  # Maximum
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.max")
  )
  expect_is(ipmat, "MaicInput")
  expect_equal(sum(ipmat$excluded),
               sum(stackloss$Air.Flow > mtch.targets[["Air.Flow.max"]]))
  expect_equal(dim(ipmat$input.matrix), c(sum(!ipmat$excluded), 0))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 0)
  
  # SD
  ipmat.sd <- createMAICInput(index = stackloss,
                              target = mtch.targets,
                              dictionary = mtch.dict,
                              matching.variables = c("airflow.sd")
  )
  expect_is(ipmat.sd, "MaicInput")
  expect_equal(dim(ipmat.sd$input.matrix), c(21, 1))
  expect_match(colnames(ipmat.sd$input.matrix), "airflow.sd")
  expect_true(!(any(ipmat.sd$excluded)))
  expect_equal(ipmat.sd$n.adjustments, 1)
  expect_equal(ipmat.sd$n.matches, 1)
  expect_equal_to_reference(ipmat.sd$input.matrix,
                            "ipmat.sd.input.matrix.rds")
  
  # Var
  ipmat.var <- createMAICInput(index = stackloss,
                               target = mtch.targets,
                               dictionary = mtch.dict,
                               matching.variables = c("airflow.var")
  )
  expect_is(ipmat.var, "MaicInput")
  expect_equal(dim(ipmat.var$input.matrix), c(21, 1))
  expect_match(colnames(ipmat.var$input.matrix), "airflow.var")
  expect_true(!(any(ipmat.var$excluded)))
  expect_equal(ipmat.var$n.adjustments, 1)
  expect_equal(ipmat.var$n.matches, 1)
  expect_equal_to_reference(ipmat.var$input.matrix,
                            "ipmat.var.input.matrix.rds")
  
  expect_equal(ipmat.sd$input.matrix[,1], ipmat.var$input.matrix[,1])
  
  rm(list = c("ipmat.sd",
              "ipmat.var",
              "ipmat"))
  
  # Quantiles
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("watertemp.10tile"))
  expect_is(ipmat, "MaicInput")
  expect_equal(dim(ipmat$input.matrix), c(21, 1))
  expect_match(colnames(ipmat$input.matrix), c("watertemp.10tile"))
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 1)
  expect_equal(sum(ipmat$input.matrix>0),
               sum(stackloss$Water.Temp <= 
                     mtch.targets[["Water.Temp.decile.1"]]))
  
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("watertemp.90tile"))
  expect_is(ipmat, "MaicInput")
  expect_equal(dim(ipmat$input.matrix), c(21, 1))
  expect_match(colnames(ipmat$input.matrix), c("watertemp.90tile"))
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 1)
  expect_equal(sum(ipmat$input.matrix>0),
               sum(stackloss$Water.Temp < 
                     mtch.targets[["Water.Temp.decile.9"]]))
  
  # Median
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("watertemp.median"))
  expect_is(ipmat, "MaicInput")
  expect_equal(dim(ipmat$input.matrix), c(21, 1))
  expect_match(colnames(ipmat$input.matrix), c("watertemp.median"))
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 1)
  expect_equal(ipmat$n.matches, 1)
  expect_equal(sum(ipmat$input.matrix>0),
               sum(stackloss$Water.Temp < 
                     mtch.targets[["Water.Temp.median"]]))
})

test_that("Sensible errors are thrown if the input constructor is fed bad input", {
  # Use the datasets::stackloss dataset for these tests
  data("stackloss")
  stackloss$match.conc.lt.90 <- ifelse(stackloss$Acid.Conc. < 90, 1, 0)
  
  mtch.targets <- list(
    "Air.Flow.mean" = 60,
    "Air.Flow.min" = 55,
    "Air.Flow.max" = 90,
    "Air.Flow.sd" = 9,
    "Air.Flow.var" = 81,
    "Proportion.Acid.Conc.lt.90" = 0.7,
    "Water.Temp.decile.1" = 18,
    "Water.Temp.decile.9" = 25,
    "Water.Temp.median" = 22
  )
  
  mtch.dict <- data.frame(match.id = c(
    "airflow.mean", "airflow.min", "airflow.max", "airflow.sd", 
    "airflow.var", "acidconc.prop",
    "watertemp.10tile", "watertemp.90tile", "watertemp.median"
  ),
  match.type = c(
    "mean", "min", "max", "sd",
    "var", "proportion",
    "quantile.1", "quantile.90", "median"
  ),
  target.variable = c(
    "Air.Flow.mean", "Air.Flow.min", "Air.Flow.max", "Air.Flow.sd",
    "Air.Flow.var", "Proportion.Acid.Conc.lt.90",
    "Water.Temp.decile.1", "Water.Temp.decile.9", "Water.Temp.median"
  ),
  index.variable = c(
    "Air.Flow", "Air.Flow", "Air.Flow", "Air.Flow", 
    "Air.Flow", "match.conc.lt.90",
    "Water.Temp", "Water.Temp", "Water.Temp"
  ),
  supplementary.target.variable = c(
    "", "", "", "Air.Flow.mean",
    "Air.Flow.mean", "",
    "", "", ""
  ),
  stringsAsFactors = FALSE
  )
  
  # Check that things work for ordinary creation
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean",
                                                  "airflow.sd",
                                                  "airflow.min",
                                                  "airflow.max",
                                                  "acidconc.prop",
                                                  "watertemp.10tile",
                                                  "watertemp.90tile",
                                                  "watertemp.median")
  )
  expect_is(ipmat, "MaicInput")
  
  # Check that things work for truncated variables
  ipmat <- createMAICInput(ind = stackloss,
                           tar = mtch.targets,
                           dic = mtch.dict,
                           mat = c("airflow.mean",
                                                  "airflow.sd",
                                                  "airflow.min",
                                                  "airflow.max",
                                                  "acidconc.prop",
                                                  "watertemp.10tile",
                                                  "watertemp.90tile",
                                                  "watertemp.median")
  )
  expect_is(ipmat, "MaicInput")
  
  # Check that things work if there are no matches
  ipmat <- createMAICInput(ind = stackloss,
                           tar = mtch.targets,
                           dic = mtch.dict,
                           mat = c()
  )
  expect_is(ipmat, "MaicInput")
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 0)
  expect_equal(ipmat$n.matches, 0)
  
  # Check failure if there are vital colnames missing
  d2 <- mtch.dict
  d2 <- d2[, -which(colnames(d2) %in% c("match.id")), drop = FALSE]
  expect_error(createMAICInput(ind = stackloss,
                           tar = mtch.targets,
                           dic = d2,
                           mat = c("airflow.mean",
                                   "airflow.sd",
                                   "airflow.min",
                                   "airflow.max",
                                   "acidconc.prop",
                                   "watertemp.10tile",
                                   "watertemp.90tile",
                                   "watertemp.median")))
  
  # Check easy passthrough if there are no matches
  ipmat <- createMAICInput(ind = stackloss,
                               tar = mtch.targets,
                               dic = d2,
                               mat = c())
  expect_is(ipmat, "MaicInput")
  expect_true(!(any(ipmat$excluded)))
  expect_equal(ipmat$n.adjustments, 0)
  expect_equal(ipmat$n.matches, 0)
  
  # Check failure if there are aux colnames missing
  d2 <- mtch.dict
  d2 <- d2[, -which(colnames(d2) %in% c("supplementary.target.variable")),
           drop = FALSE]
  expect_error(createMAICInput(ind = stackloss,
                               tar = mtch.targets,
                               dic = d2,
                               mat = c("airflow.mean",
                                       "airflow.sd",
                                       "airflow.min",
                                       "airflow.max",
                                       "acidconc.prop",
                                       "watertemp.10tile",
                                       "watertemp.90tile",
                                       "watertemp.median")))
  
  # Check OK if we take out the SD match
  ipmat <- createMAICInput(ind = stackloss,
                           tar = mtch.targets,
                           dic = d2,
                           mat = c("airflow.mean",
                                   "airflow.min",
                                   "airflow.max",
                                   "acidconc.prop",
                                   "watertemp.10tile",
                                   "watertemp.90tile",
                                   "watertemp.median"))
  expect_is(ipmat, "MaicInput")
  
  rm(d2)
  
  # What happens if we try to set up a quantile match outside the domain
  # of our data?
  mtch.targets.2 <- mtch.targets
  mtch.targets.2[["Water.Temp.decile.1"]] <- min(stackloss$Water.Temp) - 1
  expect_error(createMAICInput(ind = stackloss,
                               tar = mtch.targets.2,
                               dic = mtch.dict,
                               mat = c("airflow.mean",
                                   "airflow.min",
                                   "airflow.max",
                                   "acidconc.prop",
                                   "watertemp.10tile",
                                   "watertemp.90tile",
                                   "watertemp.median")),
    "Cannot match quantile for watertemp.10tile, target is below minimum")
  
  mtch.targets.2 <- mtch.targets
  mtch.targets.2[["Water.Temp.decile.9"]] <- max(stackloss$Water.Temp) + 1
  expect_error(createMAICInput(ind = stackloss,
                               tar = mtch.targets.2,
                               dic = mtch.dict,
                               mat = c("airflow.mean",
                                       "airflow.min",
                                       "airflow.max",
                                       "acidconc.prop",
                                       "watertemp.10tile",
                                       "watertemp.90tile",
                                       "watertemp.median")),
    "Cannot match quantile for watertemp.90tile, target is above maximum")
  
  # Median match outside domain
  mtch.targets.2 <- mtch.targets
  mtch.targets.2[["Water.Temp.median"]] <- min(stackloss$Water.Temp) - 1
  expect_error(createMAICInput(ind = stackloss,
                               tar = mtch.targets.2,
                               dic = mtch.dict,
                               mat = c("airflow.mean",
                                       "airflow.min",
                                       "airflow.max",
                                       "acidconc.prop",
                                       "watertemp.10tile",
                                       "watertemp.90tile",
                                       "watertemp.median")),
    "Cannot match median for watertemp.median, target is below minimum")
  
  mtch.targets.2 <- mtch.targets
  mtch.targets.2[["Water.Temp.median"]] <- max(stackloss$Water.Temp) + 1
  expect_error(createMAICInput(ind = stackloss,
                               tar = mtch.targets.2,
                               dic = mtch.dict,
                               mat = c("airflow.mean",
                                       "airflow.min",
                                       "airflow.max",
                                       "acidconc.prop",
                                       "watertemp.10tile",
                                       "watertemp.90tile",
                                       "watertemp.median")),
    "Cannot match median for watertemp.median, target is above maximum")
  
  # Data missing from index
  stackloss.2 <- stackloss[, - which(colnames(stackloss) %in% "Air.Flow")]
  expect_error(createMAICInput(index = stackloss.2,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean",
                                                  "airflow.sd",
                                                  "airflow.min",
                                                  "airflow.max",
                                                  "acidconc.prop",
                                                  "watertemp.10tile",
                                                  "watertemp.90tile",
                                                  "watertemp.median")),
    "Air.Flow not in index data")
  
  # Data missing from target
  mtch.targets.2 <- mtch.targets[-which(names(mtch.targets) == "Air.Flow.mean")]
  expect_error(createMAICInput(index = stackloss,
                  target = mtch.targets.2,
                  dictionary = mtch.dict,
                  matching.variables = c("airflow.mean",
                                         "airflow.sd",
                                         "airflow.min",
                                         "airflow.max",
                                         "acidconc.prop",
                                         "watertemp.10tile",
                                         "watertemp.90tile",
                                         "watertemp.median")),
    "Air.Flow.mean not in target row")
})
  

tests/testthat/test_maicWeighting.R
context("MAIC weighting")
library(maic)

test_that("Normal processing of correctly specified matching",{
  
  data("stackloss")
  stackloss$match.conc.lt.90 <- ifelse(stackloss$Acid.Conc. < 90, 1, 0)
  
  mtch.targets <- list(
    "Air.Flow.mean" = 65,
    "Air.Flow.min" = 55,
    "Air.Flow.max" = 90,
    "Air.Flow.sd" = 9,
    "Air.Flow.var" = 81,
    "Proportion.Acid.Conc.lt.90" = 0.7,
    "Water.Temp.decile.1" = 18,
    "Water.Temp.decile.9" = 25,
    "Water.Temp.median" = 22,
    "Water.Temp.percentile.5" = 18
  )
  
  mtch.dict <- data.frame(match.id = c(
    "airflow.mean", "airflow.min", "airflow.max", "airflow.sd", 
    "airflow.var", "acidconc.prop",
    "watertemp.10tile", "watertemp.90tile", "watertemp.median",
    "watertemp.05tile"
  ),
  match.type = c(
    "mean", "min", "max", "sd",
    "var", "proportion",
    "quantile.1", "quantile.90", "median",
    "quantile.05"
  ),
  target.variable = c(
    "Air.Flow.mean", "Air.Flow.min", "Air.Flow.max", "Air.Flow.sd",
    "Air.Flow.var", "Proportion.Acid.Conc.lt.90",
    "Water.Temp.decile.1", "Water.Temp.decile.9", "Water.Temp.median",
    "Water.Temp.percentile.5"
  ),
  index.variable = c(
    "Air.Flow", "Air.Flow", "Air.Flow", "Air.Flow", 
    "Air.Flow", "match.conc.lt.90",
    "Water.Temp", "Water.Temp", "Water.Temp",
    "Water.Temp"
  ),
  supplementary.target.variable = c(
    "", "", "", "Air.Flow.mean",
    "Air.Flow.mean", "",
    "", "", "",
    ""
  ),
  stringsAsFactors = FALSE
  )
  
  ##### Each matching type #####
  # Mean
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean")
  )
  
  wts <- maicWeight(ipmat, control = list(reltol = 1e-12))
  expect_equal(length(wts), 21)
  
  # We expect all unique values of Air.Flow to have the same weight
  af.fct <- factor(stackloss$Air.Flow)
  same.weight <- TRUE
  for(lvl in levels(af.fct)){
    same.weight <- same.weight || length(unique(wts[af.fct==lvl])) != 1
  }
  expect_true(same.weight)
  # This is a little bit strong maybe, but this is the target:
  expect_equal(sum(wts * stackloss$Air.Flow) / sum(wts), 
               mtch.targets[["Air.Flow.mean"]],
               tolerance = 1e-6)
  
  # Proportion
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("acidconc.prop"))
  wts <- maicWeight(ipmat)
  expect_equal(length(wts),21)
  
  # We expect all unique values of Air.Flow to have the same weight
  af.fct <- factor(stackloss$match.conc.lt.90)
  same.weight <- TRUE
  for(lvl in levels(af.fct)){
    same.weight <- same.weight || length(unique(wts[af.fct==lvl])) != 1
  }
  expect_true(same.weight)
  expect_equal(sum(wts * stackloss$match.conc.lt.90) / sum(wts), 
               mtch.targets[["Proportion.Acid.Conc.lt.90"]],
               tolerance = 1e-6)
  
  # Minimum
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.min")
  )
  wts <- maicWeight(ipmat)
  expect_equal(length(wts), 21)
  expect_true(all(wts[stackloss$Air.Flow < mtch.targets[["Air.Flow.min"]]] == 0))
  expect_true(all(wts[stackloss$Air.Flow >= mtch.targets[["Air.Flow.min"]]] == 1))
  
  # Maximum
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.max")
  )
  wts <- maicWeight(ipmat)
  expect_equal(length(wts), 21)
  expect_true(all(wts[stackloss$Air.Flow > 
                        mtch.targets[["Air.Flow.max"]]] == 0))
  expect_true(all(wts[stackloss$Air.Flow <= 
                        mtch.targets[["Air.Flow.max"]]] == 1))
  
  # SD
  ipmat.sd <- createMAICInput(index = stackloss,
                              target = mtch.targets,
                              dictionary = mtch.dict,
                              matching.variables = c("airflow.mean", 
                                                     "airflow.sd")
  )
  wts.sd <- maicWeight(ipmat.sd,
                    control = list(reltol = 1e-12))
  expect_equal(length(wts.sd), 21)
  
  # We expect all unique values of Air.Flow to have the same weight
  af.fct <- factor(stackloss$Air.Flow)
  same.weight <- TRUE
  for(lvl in levels(af.fct)){
    same.weight <- same.weight || length(unique(wts.sd[af.fct==lvl])) != 1
  }
  expect_true(same.weight)
  expect_equal(sqrt(Hmisc::wtd.var(stackloss$Air.Flow,
                                   wts.sd,
                                   method = "ML")),
               mtch.targets[["Air.Flow.sd"]],
               tolerance = 1e-6)
  
  # Variance
  ipmat.var <- createMAICInput(index = stackloss,
                               target = mtch.targets,
                               dictionary = mtch.dict,
                               matching.variables = c("airflow.mean",
                                                      "airflow.var"))
  wts.var <- maicWeight(ipmat.var,
                    control = list(reltol = 1e-12))
  expect_equal(length(wts.var), 21)
  
  # We expect all unique values of Air.Flow to have the same weight
  af.fct <- factor(stackloss$Air.Flow)
  same.weight <- TRUE
  for(lvl in levels(af.fct)){
    same.weight <- same.weight || length(unique(wts.var[af.fct==lvl])) != 1
  }
  expect_true(same.weight)
  expect_equal(Hmisc::wtd.var(stackloss$Air.Flow,
                                   wts.var,
                                   method = "ML"),
               mtch.targets[["Air.Flow.var"]],
               tolerance = 1e-6)
  
  expect_true(all(wts.sd == wts.var))
  
  rm(ipmat, ipmat.sd, ipmat.var, wts, wts.sd, wts.var)
  
  # Quantiles
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = "watertemp.10tile")
  wts <- maicWeight(ipmat,
                    control = list(reltol = 1e-12))
  
  expect_equal(length(wts), 21)
  expect_true(length(unique(wts[stackloss$Water.Temp < 
                           mtch.targets[["Water.Temp.decile.1"]]])) == 1)
  expect_true(length(unique(wts[stackloss$Water.Temp > 
                           mtch.targets[["Water.Temp.decile.1"]]])) == 1)
  # Have to have quite a high tolerance here
  expect_equal(as.numeric(Hmisc::wtd.quantile(stackloss$Water.Temp,
                                   wts * 100000, # Overcome issue with Hmisc and non-integer weights
                                   probs = 0.1)),
               mtch.targets[["Water.Temp.decile.1"]],
               tolerance = 1e-2)
  
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = "watertemp.90tile")
  wts <- maicWeight(ipmat,
                    control = list(reltol = 1e-12))
  
  expect_equal(length(wts), 21)
  expect_true(length(unique(wts[stackloss$Water.Temp <
                           mtch.targets[["Water.Temp.decile.9"]]])) == 1)
  expect_true(length(unique(wts[stackloss$Water.Temp > 
                           mtch.targets[["Water.Temp.decile.9"]]])) == 1)
  expect_equal(as.numeric(Hmisc::wtd.quantile(stackloss$Water.Temp,
                                              wts * 100000, # Overcome issue with Hmisc and non-integer weights
                                              probs = 0.9)),
               mtch.targets[["Water.Temp.decile.9"]],
               tolerance = 1e-2)
  
  # Leading-zero quantile
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = "watertemp.05tile")
  wts <- maicWeight(ipmat,
                    control = list(reltol = 1e-12))
  
  expect_equal(length(wts), 21)
  expect_true(length(unique(wts[stackloss$Water.Temp < 
                                  mtch.targets[["Water.Temp.percentile.5"]]])) == 1)
  expect_true(length(unique(wts[stackloss$Water.Temp > 
                                  mtch.targets[["Water.Temp.percentile.5"]]])) == 1)
  # Have to have quite a high tolerance here
  expect_equal(as.numeric(Hmisc::wtd.quantile(stackloss$Water.Temp,
                                              wts * 100000, # Overcome issue with Hmisc and non-integer weights
                                              probs = 0.05)),
               mtch.targets[["Water.Temp.percentile.5"]],
               tolerance = 1e-2)
})

test_that("Highly stressed matching",{
  
  data("stackloss")
  stackloss$match.conc.lt.90 <- ifelse(stackloss$Acid.Conc. < 90, 1, 0)
  
  mtch.dict <- data.frame(match.id = c(
    "airflow.mean", "airflow.min", "airflow.max", "airflow.sd", 
    "airflow.var", "acidconc.prop",
    "watertemp.10tile", "watertemp.90tile", "watertemp.median"
  ),
  match.type = c(
    "mean", "min", "max", "sd",
    "var", "proportion",
    "quantile.1", "quantile.90", "median"
  ),
  target.variable = c(
    "Air.Flow.mean", "Air.Flow.min", "Air.Flow.max", "Air.Flow.sd",
    "Air.Flow.var", "Proportion.Acid.Conc.lt.90",
    "Water.Temp.decile.1", "Water.Temp.decile.9", "Water.Temp.median"
  ),
  index.variable = c(
    "Air.Flow", "Air.Flow", "Air.Flow", "Air.Flow", 
    "Air.Flow", "match.conc.lt.90",
    "Water.Temp", "Water.Temp", "Water.Temp"
  ),
  supplementary.target.variable = c(
    "", "", "", "Air.Flow.mean",
    "Air.Flow.mean", "",
    "", "", ""
  ),
  stringsAsFactors = FALSE
  )
  
  ##### Subgrouping #####
  mtch.targets <- list(
    "Proportion.Acid.Conc.lt.90" = 0
  )
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("acidconc.prop")
  )
  wts <- maicWeight(ipmat, control = list(reltol = 1e-12))
  expect_equal(sum(wts!=0), sum(stackloss$match.conc.lt.90 == 0))
  expect_true(all(wts[stackloss$match.conc.lt.90 == 1] == 0))
  
  mtch.targets <- list(
    "Proportion.Acid.Conc.lt.90" = 1
  )
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("acidconc.prop")
  )
  wts <- maicWeight(ipmat, control = list(reltol = 1e-12))
  expect_equal(sum(wts!=0), sum(stackloss$match.conc.lt.90 == 1))
  expect_true(all(wts[stackloss$match.conc.lt.90 == 0] == 0))
  
  ##### Force errors #####
  # Mean exceeding domain
  mtch.targets <- list(
    "Air.Flow.mean" = 85
  )
  ipmat <- createMAICInput(index = stackloss,
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean")
  )
  
  wts <- maicWeight(ipmat, control = list(reltol = 1e-12))
  expect_equal(length(wts), 21)
  expect_true(all(wts == 0))
  
  # More targets than observations
  mtch.targets <- list(
    "Air.Flow.mean" = 60,
    "Air.Flow.min" = 55,
    "Air.Flow.max" = 90,
    "Air.Flow.sd" = 9,
    "Air.Flow.var" = 81,
    "Proportion.Acid.Conc.lt.90" = 0.7,
    "Water.Temp.decile.1" = 18,
    "Water.Temp.decile.9" = 25,
    "Water.Temp.median" = 22
  )
  
  ipmat <- createMAICInput(index = stackloss[c(1,10,21), ],
                           target = mtch.targets,
                           dictionary = mtch.dict,
                           matching.variables = c("airflow.mean",
                                                  "airflow.var",
                                                  "acidconc.prop",
                                                  "watertemp.median")
                           )
  expect_warning(wts <- maicWeight(ipmat, 
                                   control = list(reltol = 1e-12,
                                          maxit = 5000)),
                 "Optimisation has not converged")
})

R/examples/maic.example.R
target <- c("Air.Flow" = 60,
           "Water.Temp" = 21,
           "Prop.Acid.Conc.LT.90" = 0.7,
           "min.air.flow" = 55)

stackloss$match.conc.lt.90 <- 
  ifelse(stackloss$Acid.Conc. < 90, 1, 0)

dict <- data.frame(
  "match.id" = 
    c("airflow", "watertemp", 
      "acidconc", "min.airflow"),
  "target.variable" = 
    c("Air.Flow", "Water.Temp",
      "Prop.Acid.Conc.LT.90", "min.air.flow"),
  "index.variable" = 
    c("Air.Flow", "Water.Temp",
      "match.conc.lt.90", "Air.Flow"),
  "match.type" = 
    c("mean", "mean", "proportion", "min"),
  stringsAsFactors = FALSE)

ipmat <- createMAICInput(
  index = stackloss,
  target = target,
  dictionary = dict,
  matching.variables = 
    c("airflow", "watertemp", 
      "acidconc", "min.airflow"))

wts <- maicWeight(ipmat)

rcv <- reportCovariates(
  stackloss, target, dict, 
  matching.variables = 
    c("airflow", "watertemp", 
      "acidconc", "min.airflow"),
  wts)
